### PR TITLE
`FileAppender` GZIP support

### LINKED ISSUE #759: Provide FileAppender pipeline element option to compress output [Enhancement]
In some circumstances, users can export significant amounts of data via file Appenders. To minimise  storage cost and external script integration, it would be useful to optionally compress (zip or gzip) the resultant output files.

This would be particularly useful for rollingFileAppenders (say a compressRolledFileName option with zip/gzip argument), although there would be no reason (aside from complexity to implement) compressing files generated from a simple fileAppender.

### PR DESCRIPTION
**Pull request by p-kimberley**
Provide an option to apply GZIP compression to `FileAppender` output files. Closes #759.

### CODE REVIEW

File: stroom-pipeline/src/main/java/stroom/pipeline/server/writer/FileAppender.java [hunk 1]
```diff
@@ -132,6 +140,9 @@ public void setOutputPaths(final String outputPaths) {
         this.outputPaths = outputPaths.split(",");
     }
 
+    @PipelineProperty(description = "Apply GZIP compression to output files", defaultValue = "false")
+    public void setUseCompression(final boolean useCompression) { this.useCompression = useCompression; }
```
Comment (at055612) [inline stroom-pipeline/src/main/java/stroom/pipeline/server/writer/FileAppender.java:144]: public void setUseCompression(final boolean useCompression) {
        this.useCompression = useCompression;
    }

File: stroom-pipeline/src/main/java/stroom/pipeline/server/writer/FileAppender.java [hunk 2]
```diff
@@ -106,7 +108,13 @@ protected OutputStream createOutputStream() throws IOException {
             }
 
             // Get a writer for the new lock file.
-            byteCountOutputStream = new ByteCountOutputStream(new BufferedOutputStream(Files.newOutputStream(lockFile)));
+            if (useCompression) {
+                byteCountOutputStream = new ByteCountOutputStream(new GZIPOutputStream(new BufferedOutputStream(Files.newOutputStream(lockFile))));
+            }
+            else {
```
Comment (at055612) [inline stroom-pipeline/src/main/java/stroom/pipeline/server/writer/FileAppender.java:114]: } else {

File: stroom-pipeline/src/main/java/stroom/pipeline/server/writer/FileAppender.java [hunk 3]
```diff
@@ -106,7 +108,13 @@ protected OutputStream createOutputStream() throws IOException {
             }
 
             // Get a writer for the new lock file.
-            byteCountOutputStream = new ByteCountOutputStream(new BufferedOutputStream(Files.newOutputStream(lockFile)));
+            if (useCompression) {
+                byteCountOutputStream = new ByteCountOutputStream(new GZIPOutputStream(new BufferedOutputStream(Files.newOutputStream(lockFile))));
```
Comment (at055612) [inline stroom-pipeline/src/main/java/stroom/pipeline/server/writer/FileAppender.java:112]: Possibly too long.  We have a hard limit of 120chrs in v7

**Review by at055612 — CHANGES_REQUESTED** [review]

**Review by p-kimberley — COMMENTED** [review]
Thanks @at055612

**Comment by stroomdev66** [comment]
What about `RollingFileAppender`?

**Comment by p-kimberley** [comment]
@stroomdev66 Fair enough :) Quite a simple change as well, will get to it shortly.

Won't back-port changes to 5.5 though - unless there's an urgent need for them.

**Pull request merged**